### PR TITLE
LayerNorm Fake FP16 Op debug

### PR DESCRIPTION
**Pull request by venkatacrc**
LayerNorm Fake FP16 Op debug. 
still seeing output mismatches.

### CODE REVIEW

File: caffe2/contrib/fakelowp/layernorm_fp16_fake_op.cc [hunk 1]
```diff
@@ -1,8 +1,8 @@
-#include "layernorm_fp16_fake_op.h"
+﻿#include "layernorm_fp16_fake_op.h"
```
Comment (hyuen) [inline caffe2/contrib/fakelowp/layernorm_fp16_fake_op.cc:1]: what is the change here? something with your editor?

File: caffe2/contrib/fakelowp/layernorm_fp16_fake_op.h [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#pragma once
+﻿#pragma once
```
Comment (hyuen) [inline caffe2/contrib/fakelowp/layernorm_fp16_fake_op.h:1]: same

File: caffe2/contrib/fakelowp/layernorm_fp16_fake_op.h [hunk 3]
```diff
@@ -75,7 +79,49 @@ class LayerNormFakeFp16Op final : public Operator<Context> {
         mean_data,
         sigma_data,
         &context_);
+//    ComputeSigmaAndFusedParams<T>(
```
Comment (hyuen) [inline caffe2/contrib/fakelowp/layernorm_fp16_fake_op.h:82]: remove

File: caffe2/contrib/fakelowp/layernorm_fp16_fake_op.h [hunk 4]
```diff
@@ -75,7 +79,49 @@ class LayerNormFakeFp16Op final : public Operator<Context> {
...
+    ConstEigenVectorArrayMap<T> var_arr(sigma_data, M);
+    EigenVectorArrayMap<T> sigma_arr(sigma_data, M);
+    sigma_arr = var_arr + static_cast<T>(epsilon_);
+    math::Rsqrt<T, CPUContext>(M, sigma_data, scale_data, &context_);
+    math::Mul<T, CPUContext>(M, scale_data, sigma_data, sigma_data, &context_);
+    EigenVectorArrayMap<T>(bias_data, M) =
+        -ConstEigenVectorArrayMap<T>(scale_data, M) *
+        ConstEigenVectorArrayMap<T>(mean_data, M);
 
+//    LayerNormForward<T>(
```
Comment (hyuen) [inline caffe2/contrib/fakelowp/layernorm_fp16_fake_op.h:93]: were you planning to abstract this into this function? also with ComputeSigmaAndFusedParams? i think it is a good idea to do so

File: caffe2/contrib/fakelowp/test/test_layernorm_nnpi_fp16.py [hunk 5]
```diff
@@ -23,15 +26,17 @@
...
 # Test the lowered LayerNorm op
-class LayerNorm(TestCase):
-    def _test_layernorm(self):
-        size = 3
-        input_channels = 2
-        batch_size = 4
-        seed = int(time.time())
+class LayerNorm(serial.SerializedTestCase):
+    @given(seed=st.integers(0, 65535),
+        size = st.integers(2,3),
```
Comment (hyuen) [inline caffe2/contrib/fakelowp/test/test_layernorm_nnpi_fp16.py:31]: increase the ranges for these arguments

File: caffe2/contrib/fakelowp/test/test_layernorm_nnpi_fp16.py [hunk 6]
```diff
@@ -23,15 +26,17 @@
...
+        input_channels=st.integers(1, 2),
+        batch_size=st.integers(2,4),
+        epsilon=st.floats(min_value=1e-4, max_value=1e-2))
+    @settings(max_examples=100)
+    def _test_layernorm(self, seed, size, input_channels, batch_size, epsilon):
```
Comment (hyuen) [inline caffe2/contrib/fakelowp/test/test_layernorm_nnpi_fp16.py:36]: is this running? remove the _

**Review by hyuen — COMMENTED** [review]
please address the comments

File: caffe2/contrib/fakelowp/layernorm_fp16_fake_op.cc (hunk 1 above)
Comment (venkatacrc) [reply caffe2/contrib/fakelowp/layernorm_fp16_fake_op.cc:1]: I think so. I will look into it.

File: caffe2/contrib/fakelowp/layernorm_fp16_fake_op.h (hunk 4 above)
Comment (venkatacrc) [reply caffe2/contrib/fakelowp/layernorm_fp16_fake_op.h:93]: I will abstract this into a function.

**Review by hyuen — APPROVED** [review]

**Pull request closed without merging**